src/lib.rs

```diff
@@ -0,0 +1,289 @@
+        }
+
+        // Safe because the DataInit trait asserts any data is valid for this type, and we ensured
+        // the size of the pointer's buffer is the correct size. The `align_to` method ensures that
+        // we don't have any unaligned references. This aliases a pointer, but because the pointer
+        // is from a mut slice reference, we borrow the passed in mutable reference. Finally, the
+        // reference returned can not outlive data because they have equal implicit lifetime
+        // constraints.
+        match unsafe { data.align_to_mut::<Self>() } {
+            ([], [mid], []) => Some(mid),
+            _ => None,
+        }
+    }
+
+    /// Converts a reference to `self` into a slice of bytes.
+    ///
+    /// The value of `self` is not copied. Instead, the slice is made from a reference to `self`.
+    /// The value of bytes in the returned slice will depend on the representation of the type in
+    /// memory, and may change in an unstable fashion.
+    fn as_slice(&self) -> &[u8] {
+        // Safe because the entire size of self is accessible as bytes because the trait guarantees
+        // it. The lifetime of the returned slice is the same as the passed reference, so that no
+        // dangling pointers will result from this pointer alias.
+        unsafe { from_raw_parts(self as *const Self as *const u8, size_of::<Self>()) }
+    }
+
+    /// Converts a mutable reference to `self` into a mutable slice of bytes.
+    ///
+    /// Because the slice is made from a reference to `self`, mutations to the returned slice are
+    /// immediately reflected in `self`. The value of bytes in the returned slice will depend on
+    /// the representation of the type in memory, and may change in an unstable fashion.
+    fn as_mut_slice(&mut self) -> &mut [u8] {
+        // Safe because the entire size of self is accessible as bytes because the trait guarantees
+        // it. The trait also guarantees that any combination of bytes is valid for this type, so
+        // modifying them in the form of a byte slice is valid. The lifetime of the returned slice
+        // is the same as the passed reference, so that no dangling pointers will result from this
+        // pointer alias. Although this does alias a mutable pointer, we do so by exclusively
+        // borrowing the given mutable reference.
+        unsafe { from_raw_parts_mut(self as *mut Self as *mut u8, size_of::<Self>()) }
+    }
+}
+
+/// A container to host byte and access its content.
+///
+/// Candidates implement this trait include:
+/// - anonymous memory areas
+/// - mmapped memory areas
+/// - data files
+/// - a proxy to access memory on remote
+pub trait Bytes<A> {
+    /// Associated error codes
+    type E;
+
+    /// Writes a slice to the region at the specified address.
+    /// Returns the number of bytes written. The number of bytes written can
+    /// be less than the length of the slice if there isn't enough room in the
+    /// region.
```

Comment:
Yes, the only difference is that write_slice() verifies that the whole buffer has been write successfully.
But it's hard to enumerate possible error cases of write() because it depends on the backend implementation.